### ISSUE TITLE
eBPF control path enters epochs with hard CPU affinity but without priority boost

### ISSUE THREAD

**Issue by mtfriesen** [microsoft/ebpf-for-windows#2126]
### Describe the bug

XDP stress tests are hitting control path watchdogs  which are triggered when a thread calling an eBPF or XDP API takes an exceptionally long amount of time, on the order of 10 seconds.

The scenario boils down to:

1. A user mode thread running at default, unaffinitized priority calls an eBPF API.
2. The eBPF driver enters an eBPF epoch while it happens to be running on a highly contended CPU.
3. The eBPF driver sets the thread affinity to the current CPU.
4. The scheduler preempts the eBPF thread for a higher priority thread and cannot schedule onto a less busy CPU. In this case, the higher priority thread is the networking data path, but it could be any thread. For a bit of context, the newest NDIS data paths run by default with a base priority of 10, and the default thread base priority for Windows is 8. 

Exacerbating this scenario is driver verifier performing a working set trim in the context of eBPF. However, the very small CPU usage time in the `!thread` output below indicates this thread has been starved rather than taking an exceptionally long time in the driver verifier routine.

Since eBPF must pin the thread to a specific CPU within an epoch, and presumably this control path must be within an epoch, one possible mitigation is to raise the thread priority before setting the thread affinity in the `ebpf_epoch_enter` routine. If eBPF were to do this, the next question is how much to raise the priority.

### OS information

17763.1.amd64fre.rs5_release.180914-1434

### Steps taken to reproduce bug

Run XDP stress tests with eBPF integration.

### Expected behavior

eBPF APIs should return as quickly as feasible without preempting even more critical work on the system.

### Actual outcome

eBPF API takes a very long time to return on a (partially) busy system.

### Additional details

```
3: kd> !thread ffff960f`e020c0c0
THREAD ffff960fe020c0c0  Cid 1760.19fc  Teb: 000000957fea6000 Win32Thread: 0000000000000000 READY on processor 2
IRP List:
    ffff960fdff24e50: (0006,01a8) Flags: 40060070  Mdl: 00000000
Not impersonating
DeviceMap                 ffffb8036b32af60
Owning Process            ffff960fdfb880c0       Image:         spinxsk.exe
Attached Process          N/A            Image:         N/A
Wait Start TickCount      564892         Ticks: 11 (0:00:00:00.171)
Context Switch Count      143            IdealProcessor: 2             
UserTime                  00:00:00.000
KernelTime                00:00:00.015
Win32 Start Address spinxsk!ILT+985(XskFuzzerWorkerFn) (0x00007ff6849eb3de)
Stack Init fffffe02eaef3c10 Current fffffe02eaef2680
Base fffffe02eaef4000 Limit fffffe02eaeed000 Call 0000000000000000
Priority 8 BasePriority 8 PriorityDecrement 0 IoPriority 2 PagePriority 5
3: kd> kn
  *** Stack trace for last set context - .thread/.cxr resets it
 # Child-SP          RetAddr               Call Site
00 fffffe02`eaef26c0 fffff802`2cc6741c     nt!KxDispatchInterrupt+0x138 [minkernel\ntos\ke\amd64\ctxswap.asm @ 394] 
01 fffffe02`eaef2800 fffff802`2cafe6b1     nt!KiDpcInterrupt+0x2dc [minkernel\ntos\ke\amd64\dpcint.asm @ 136] 
02 fffffe02`eaef2990 fffff802`2cb10816     nt!MiUnlockWorkingSetShared+0xb1 [minkernel\ntos\mm\wslock.c @ 854] 
03 (Inline Function) --------`--------     nt!MiReleaseWalkLocks+0x2b [minkernel\ntos\mm\wslist.c @ 9640] 
04 (Inline Function) --------`--------     nt!MiYieldPageTableWalk+0x54 [minkernel\ntos\mm\wslist.c @ 8927] 
05 fffffe02`eaef29c0 fffff802`2cb0fbfa     nt!MiWalkPageTablesRecursively+0x8b6 [minkernel\ntos\mm\wslist.c @ 9985] 
06 fffffe02`eaef2aa0 fffff802`2cbe7913     nt!MiWalkPageTables+0x1da [minkernel\ntos\mm\wslist.c @ 10907] 
07 fffffe02`eaef2ba0 fffff802`2cd58178     nt!MiEmptyWorkingSetInitiate+0x103 [minkernel\ntos\mm\wslist.c @ 6560] 
08 (Inline Function) --------`--------     nt!MiEmptyWorkingSet+0x11 [minkernel\ntos\mm\wslist.c @ 6607] 
09 fffffe02`eaef2d70 fffff802`2cd58d4e     nt!MiEmptyTargetedWorkingSet+0xc0 [minkernel\ntos\mm\wsmanage.c @ 9717] 
0a fffffe02`eaef2dc0 fffff802`2d3cbc72     nt!MiTrimAllSystemPagableMemory+0xde [minkernel\ntos\mm\wsmanage.c @ 9869] 
0b fffffe02`eaef2e10 fffff802`2d3e2110     nt!MmVerifierTrimMemory+0x6a [minkernel\ntos\mm\verifier.c @ 106] 
0c fffffe02`eaef2e40 fffff802`2d3e1d9e     nt!ViKeRaiseIrqlSanityChecks+0xc8 [minkernel\ntos\verifier\vfke.c @ 498] 
0d fffffe02`eaef2e80 fffff802`2d3e0892     nt!ViKeAcquireSpinLockRaiseToDpcCommon+0x2a [minkernel\ntos\verifier\vfke.c @ 1528] 
0e fffffe02`eaef2eb0 fffff808`019b5f08     nt!VerifierKeAcquireSpinLockRaiseToDpc+0x12 [minkernel\ntos\verifier\vfke.c @ 1588] 
0f (Inline Function) --------`--------     eBPFCore!ebpf_lock_lock+0x9 [D:\ebpf-for-windows\libs\platform\kernel\ebpf_platform_kernel.c @ 381] 
10 fffffe02`eaef2ef0 fffff808`019a5120     eBPFCore!ebpf_link_attach_program+0x1fc [D:\ebpf-for-windows\libs\execution_context\ebpf_link.c @ 264] 
11 fffffe02`eaef3000 fffff808`019a2978     eBPFCore!_ebpf_core_protocol_link_program+0x100 [D:\ebpf-for-windows\libs\execution_context\ebpf_core.c @ 1138] 
12 (Inline Function) --------`--------     eBPFCore!ebpf_core_invoke_protocol_handler+0x1ea [D:\ebpf-for-windows\libs\execution_context\ebpf_core.c @ 2329] 
13 fffffe02`eaef30c0 fffff808`00de9a55     eBPFCore!_ebpf_driver_io_device_control+0x748 [D:\ebpf-for-windows\ebpfcore\ebpf_drv.c @ 333] 
14 (Inline Function) --------`--------     Wdf01000!FxIoQueueIoDeviceControl::Invoke+0x42 [minkernel\wdf\framework\shared\inc\private\common\fxioqueuecallbacks.hpp @ 226] 
15 fffffe02`eaef33e0 fffff808`00de9357     Wdf01000!FxIoQueue::DispatchRequestToDriver+0x225 [minkernel\wdf\framework\shared\irphandlers\io\fxioqueue.cpp @ 3325] 
16 fffffe02`eaef3480 fffff808`00de75a2     Wdf01000!FxIoQueue::DispatchEvents+0x617 [minkernel\wdf\framework\shared\irphandlers\io\fxioqueue.cpp @ 3125] 
17 (Inline Function) --------`--------     Wdf01000!FxIoQueue::QueueRequest+0x92 [minkernel\wdf\framework\shared\irphandlers\io\fxioqueue.cpp @ 2367] 
18 (Inline Function) --------`--------     Wdf01000!FxPkgIo::DispatchStep2+0x4c2 [minkernel\wdf\framework\shared\irphandlers\io\fxpkgio.cpp @ 469] 
19 fffffe02`eaef3560 fffff808`00de704d     Wdf01000!FxPkgIo::DispatchStep1+0x542 [minkernel\wdf\framework\shared\irphandlers\io\fxpkgio.cpp @ 324] 
1a fffffe02`eaef3620 fffff808`00de1b73     Wdf01000!FxPkgIo::Dispatch+0x5d [minkernel\wdf\framework\shared\irphandlers\io\fxpkgio.cpp @ 119] 
1b (Inline Function) --------`--------     Wdf01000!DispatchWorker+0x9e [minkernel\wdf\framework\shared\core\fxdevice.cpp @ 1572] 
1c (Inline Function) --------`--------     Wdf01000!FxDevice::Dispatch+0xbc [minkernel\wdf\framework\shared\core\fxdevice.cpp @ 1586] 
1d fffffe02`eaef3680 fffff802`2cc1763a     Wdf01000!FxDevice::DispatchWithLock+0x113 [minkernel\wdf\framework\shared\core\fxdevice.cpp @ 1430] 
1e fffffe02`eaef36e0 fffff802`2d3caf49     nt!IopfCallDriver+0x56 [minkernel\ntos\io\iomgr\iomgr.h @ 3384] 
1f fffffe02`eaef3720 fffff802`2cc8b4c7     nt!IovCallDriver+0x275 [minkernel\ntos\io\iomgr\ioverifier.c @ 589] 
20 fffffe02`eaef3760 fffff802`2d082a41     nt!IofCallDriver+0x1b74f7 [minkernel\ntos\io\iomgr\iosubs.c @ 3152] 
21 (Inline Function) --------`--------     nt!IoCallDriverWithTracing+0x2b [minkernel\ntos\io\iomgr\iomgr.h @ 1093] 
22 (Inline Function) --------`--------     nt!IopCallDriverReference+0xbd [minkernel\ntos\io\iomgr\iomgr.h @ 4221] 
23 fffffe02`eaef37a0 fffff802`2d0827b1     nt!IopSynchronousServiceTail+0x1b1 [minkernel\ntos\io\iomgr\internal.c @ 8644] 
24 fffffe02`eaef3850 fffff802`2d082876     nt!IopXxxControlFile+0xe61 [minkernel\ntos\io\iomgr\internal.c @ 11336] 
25 fffffe02`eaef39a0 fffff802`2cc72ff5     nt!NtDeviceIoControlFile+0x56 [minkernel\ntos\io\iomgr\devctrl.c @ 110] 
26 fffffe02`eaef3a10 00007ffc`d59008d4     nt!KiSystemServiceCopyEnd+0x25 [minkernel\ntos\ke\amd64\trap.asm @ 3320] 
27 00000095`001fe7e8 00007ffc`d24b96a7     ntdll!ZwDeviceIoControlFile+0x14 [minkernel\ntdll\daytona\objfre\amd64\usrstubs.asm @ 235] 
28 00000095`001fe7f0 00007ffc`d5485470     KERNELBASE!DeviceIoControl+0x67 [minkernel\kernelbase\filehops.c @ 3496] 
29 00000095`001fe860 00007ffc`b35102df     KERNEL32!DeviceIoControlImplementation+0x80 [base\win32\client\filehops.c @ 168] 
2a (Inline Function) --------`--------     EbpfApi!Platform::DeviceIoControl+0x3c [D:\ebpf-for-windows\libs\thunk\windows\platform.cpp @ 50] 
2b (Inline Function) --------`--------     EbpfApi!invoke_ioctl+0x7a [D:\ebpf-for-windows\libs\api_common\device_helper.hpp @ 99] 
2c 00000095`001fe8b0 00000000`00000000     EbpfApi!_link_ebpf_program+0x2ef [D:\ebpf-for-windows\libs\api\ebpf_api.cpp @ 1057]
```

**Comment by Alan-Jowett** [comment]
I am not sure the issue here is necessarily eBPF. The network datapath is pinned to a CPU, running for an extended period of time at an elevated priority. This behavior is going to break things other than eBPF. 

Long running work at elevated priority is the root cause of the problem. Assume a single core system: In that scenario there is no other CPU to switch lower priority work to and lower priority threads will starve.

**Comment by mtfriesen** [comment]
I agree, so perhaps this issue is mostly one of setting expectations: systems are typically configured with a strict subset of its cores devoted to networking RSS, with the rest allocated to the app workload. The app, then, expects to make timely forward progress as long as at least some CPUs on the system are not overwhelmed. It is surprising (as a layperson) that a thread calling a BPF configuration API could find itself blocked on a critical CPU.

Once the expectation is set, the app can mitigate the issue by elevating its own control path threads.

Ultimately, yes, the data path (and all driver code paths) should ideally be scheduler-friendly, and progress is gradually being made on that front.

________________________________
